README.md
# SEO

[![npm version][npm-version-src]][npm-version-href]
[![npm downloads][npm-downloads-src]][npm-downloads-href]
[![License][license-src]][license-href]
[![Nuxt][nuxt-src]][nuxt-href]

SEO module on Nuxt3 that helps you write title's, descriptions, favicons, and Schema's for doing amazing things.

- [✨ &nbsp;Release Notes](/CHANGELOG.md)
  <!-- - [🏀 Online playground](https://stackblitz.com/github/tubstrr/seo?file=playground%2Fapp.vue) -->
  <!-- - [📖 &nbsp;Documentation](https://example.com) -->

## Features

<!-- Highlight some of the features your module provide here -->

- Easy page title and description
- One time setup favicon
- One time setup Schema markup
- Easy to extend

## Quick Setup

1. Add `@tubstrr/seo` dependency to your project

```bash
# Using pnpm
pnpm add -D @tubstrr/seo

# Using yarn
yarn add --dev @tubstrr/seo

# Using npm
npm install --save-dev @tubstrr/seo
```

2. Add `@tubstrr/seo` to the `modules` section of `nuxt.config.ts`

```js
export default defineNuxtConfig({
  modules: ["@tubstrr/seo"],
});
```

3. Do a onetime SEO setup

```js
export default defineNuxtConfig({
  seo: {
    auto: false,
    locale: "en_US",
    general: {
      tagline: "We are a fiercely independent design and development studio.",
      defaultImage: "https://example.com/og-share-image.jpg",
      social: [
        {
          type: "facebook",
          link: "https://www.facebook.com/thisismess/",
        },
        {
          type: "twitter",
          link: "https://twitter.com/thisismess",
        },
        {
          type: "instagram",
          link: "https://www.instagram.com/thisismess",
        },
        {
          type: "linkedin",
          link: "https://www.linkedin.com/company/mess-marketing",
        },
        {
          type: "youtube",
          link: "https://www.youtube.com/@MessMakesVideos",
        },
      ],
    },
    organization: {
      name: "This is Mess",
      url: "https://thisismess.com",
      logo: "https://example.com/favicons/apple-touch-icon.png",
      address: {
        streetAddress: "1824 W Grand Ave #200",
        addressLocality: "Chicago",
        addressRegion: "IL",
        postalCode: "60622",
        addressCountry: "US",
      },
      extra: [
        {
          key: "ContactPoint",
          value: {
            name: "Jon Knoll",
            url: "https://jonknoll.dev",
            email: "jknoll@thisismess.com",
            image: "https://jonknoll.dev/wp-content/uploads/2021/03/use-2.jpg",
          },
        },
        { key: "email", value: "contact@thisismess.com" },
      ],
    },
    website: {
      show: true,
      name: "Mess Marketing",
      url: "https://thisismess.com",
      search: {
        show: true,
        action: "https://thisismess.com/search?search=",
      },
    },
    favicon: {
      link: [
        {
          rel: "apple-touch-icon",
          sizes: "180x180",
          href: "https://example.com/favicons/apple-touch-icon.png",
        },
        {
          rel: "icon",
          sizes: "32x32",
          href: "https://example.com/favicons/favicon-32x32.png",
        },
        {
          rel: "icon",
          sizes: "16x16",
          href: "https://example.com/favicons/favicon-16x16.png",
        },
        {
          rel: "manifest",
          href: "https://example.com/favicons/site.webmanifest",
        },
        {
          rel: "mask-icon",
          color: "#37dbff",
          href: "https://example.com/favicons/safari-pinned-tab.svg",
        },
        {
          rel: "shortcut icon",
          href: "https://example.com/favicons/favicon.ico",
        },
      ],
      meta: [
        {
          name: "msapplication-TileColor",
          content: "#ffffff",
        },
        {
          name: "theme-color",
          content: "#ffffff",
        },
      ],
    },
  },
});
```

That's it! You can now use SEO in your Nuxt app ✨

## Documentation

### Key Concepts

#### `useSeo()` composable

The `useSeo()` composable performs all the `useHead()` manipulations, based on the `SeoModulesOptions` object.

The first thing that the composable does is merge any options passed into it with the options set by the `nuxt.config` (You can see options currently set at `useRuntimeConfig().app.seo`)

The main concept is that you'll be able to write a simple ORM to map your API's into this configuration, and then it'll just work. As an example the `webpage` schema expects these keys:

- `title`
- `description`
- `datePublished`
- `dateModified`

If none are provided, `useSeo()` will use the slug to get the title, and the `general.tagline` for the description, and not show the dates.

You can easily augment this with data from your API by mapping values like this on your `[...slug].vue`:

```js
const route = useRoute();
const {data} = useFetch('/api', {
  query: {
    slug: route.path
  }
})
const options = {
  title: data.value['title'],
  description: data.value['description']
  webpage: {
    image: data.value['social-share-image'],
    datePublished: data.value['publish-date']
    dateModified: data.value['modified-date']
  }
}

useSeo(options);
```

#### addSeo() composable
the `addSeo()` composable is essentially the same set of function calls as the `useSeo()` composable, although with one key difference.  `addSeo()` will only update the options that you pass into it, it will not merge with the base configuration, and will not create the Organization, Website, Webpage, or Breadcrumb schemas.    Will allow you to add `schemas` to the page from a block level, example being an Accordion block that needs to create an FAQ Schema.

#### Full SEO Options

These are all the options that the `useSeo()` will use. When the composable is called, it will merge the `options` object that is passed into the function, with `nuxt.config.seo` object, preferring the object passed into the function.

Here are all the options available to the `nuxt.config.seo` object:

```js
// Module options TypeScript interface definition
export interface SeoModuleOptions {
  auto: boolean;
  locale: string;
  general: {
    tagline: string,
    defaultImage: string,
    social: Array<{
      type:
        | "facebook"
        | "twitter"
        | "x"
        | "youtube"
        | "instagram"
        | "linkedin"
        | "pinterest"
        | "tumblr"
        | "reddit"
        | "other",
      link: string,
    }>,
  };
  organization: {
    show: boolean,
    type: string,
    name: string,
    use: string,
    logo: string,
    address: {
      type: string,
      streetAddress: string,
      addressLocality: string,
      addressRegion: string,
      postalCode: string,
      addressCountry: string,
      extra: Array<{
        key: string,
        value: any,
      }>,
    },
  };
  website: {
    show: boolean,
    type: string,
    name: string,
    url: string,
    search: {
      show: boolean,
      action: string,
    },
    extra: Array<{
      key: string,
      value: any,
    }>,
  };
  webpage: {
    show: boolean,
    type: string,
    url: string,
    image: string,
    datePublished: string,
    dateModified: string,
    extra: Array<{
      key: string,
      value: any,
    }>,
  };
  breadcrumbs: {
    show: boolean,
    excludedPaths: Array<string>,
  };
  schemas: Array<{
    hid: string,
    schema: object,
  }>;
  title: {
    separator: string;
    template: string;
    params: object;
  }
  favicon: {
    link: Array<{
      rel:
        | "apple-touch-icon"
        | "icon"
        | "mask-icon"
        | "manifest"
        | "shortcut icon",
      href: string,
      sizes?: "16x16" | "32x32" | "96x96" | "192x192" | "512x512",
      color?: string,
    }>,
    meta: Array<{
      name: "msapplication-TileColor" | "theme-color",
      content: string,
    }>,
  };
}
```

#### auto: true/false

The `auto` key defaults to `true`, here is the code that takes this into account.

Basic idea is that if you don't want to write your own lifecycle, we'll call `useSeo()` automatically for you server side, and then when the page has finished rendering.

```js
const options = useRuntimeConfig().app?.seo;
if (!options?.auto) return;
// Server side on initial render
useSeo();
nuxtApp.hook("page:finish", () => {
  useSeo();
});
```

#### <schema>.show

You can hide any of the default schema pieces by setting <schema>.show to `false`.

```js
export default defineNuxtConfig({
  seo {
    webpage: {
      show: false
    }
  }
})
```

```js
if (options?.<schema>?.show) {
  const <schema> = {
    hid: "defaultSeoBase<schema>Schema",
    type: "application/ld+json",
    processTemplateParams: true,
    innerHTML: {...}
  }
  // ...
  schema.push(<schema>)
}
```

#### <schema>.extra

All base schema's allow for you to extend, or override any key value pair. The `extra` key allows you to write and array of key value pairs that will be injected into the base schema.

##### Example:

This will inject both the `ContactPoint` and `email` schema piece into the base `organization` schema piece

```js
export default defineNuxtConfig({
  seo {
    organization: {
      extra: [
        {
          key: "ContactPoint",
          value: {
            name: "Jon Knoll",
            url: "https://jonknoll.dev",
            email: "jknoll@thisismess.com",
            image: "https://jonknoll.dev/wp-content/uploads/2021/03/use-2.jpg",
          },
        },
        { key: "email", value: "contact@thisismess.com" },
      ]
    }
  }
})
```

#### schemas key

There are a lot of schemas out there, and there's no way we'll be able to have them automated for every site, nor should we try! This is where the `schemas` key comes in.

Similar to the `extra` key inside the base schemas, you'll be able to add your own schema pieces by adding a script ID (hid), and the schema JSON inside the `options.schemas` key. Here is an example of an event schema that will also hide the `webpage`:

```js
const options = {
  webpage: {
    show: false
  }
  schemas: [
    {
      hid: "justinBieber",
      schema: {
        "@context": "https://schema.org",
        "@type": "Event",
        name: "Name",
        description:
          "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Cras vestibulum nulla ut nibh vulputate egestas.",
        image: "https://example.com/images/jb.jpg",
        startDate: "2023-12-08T08:00",
        endDate: "2023-12-08T12:00",
        eventStatus: "https://schema.org/EventScheduled",
        eventAttendanceMode: "https://schema.org/OfflineEventAttendanceMode",
        location: {
          "@type": "Place",
          name: "Mess",
          address: {
            "@type": "PostalAddress",
            streetAddress: "1234 W Street St",
            addressLocality: "Chicago",
            addressRegion: "IL",
            postalCode: "60622",
            addressCountry: "US",
          },
        },
        performer: {
          "@type": "MusicGroup",
          name: "Justin Bieber",
        },
        offers: {
          "@type": "Offer",
          name: "General Admission",
          price: "50",
          priceCurrency: "",
          validFrom: "2023-12-04",
          url: "https://example.com/events/jb",
          availability: "https://schema.org/InStock",
        },
      },
    },
  ],
};

useSeo(options)
```

#### title key

The title key, is a simple way to augment [unhead](https://unhead.unjs.io/usage/composables/use-head)'s [`titleTemplate`](https://unhead.unjs.io/usage/guides/title-template), and [`templateParams`](https://unhead.unjs.io/usage/guides/template-params)

Here are the default options that are set for `useSeo()`:

```js
// Add default title params
options.titles.params = {
  site: {
    name: options?.website?.name,
  },
  separator: options.titles.separator,
  tagline: options.general.tagline,
};
```

#### favicon key....

There isn't really a way to trim this down. You can use a [favicon generator](https://realfavicongenerator.net/), or you can just use a simple `.png`, it's up to you! Inside the `favicon` key there is support for the correct `link`s and `meta` tags.

Here is the simplest example:

```js
seo: {
  favicon: {
    link: [
      {
        rel: "icon",
        sizes: "32x32",
        href: "https://example.com/favicons/favicon-32x32.png",
      },
    ];
  }
}
```

Here is a full setup:

```js
seo: {
  favicon: {
      link: [
        {
          rel: "apple-touch-icon",
          sizes: "180x180",
          href: "https://example.com/favicons/apple-touch-icon.png",
        },
        {
          rel: "icon",
          sizes: "32x32",
          href: "https://example.com/favicons/favicon-32x32.png",
        },
        {
          rel: "icon",
          sizes: "16x16",
          href: "https://example.com/favicons/favicon-16x16.png",
        },
        {
          rel: "manifest",
          href: "https://example.com/favicons/site.webmanifest",
        },
        {
          rel: "mask-icon",
          color: "#37dbff",
          href: "https://example.com/favicons/safari-pinned-tab.svg",
        },
        {
          rel: "shortcut icon",
          href: "https://example.com/favicons/favicon.ico",
        },
      ],
      meta: [
        {
          name: "msapplication-TileColor",
          content: "#ffffff",
        },
        {
          name: "theme-color",
          content: "#ffffff",
        },
      ],
    }
}
```

## Development

```bash
# Install dependencies
npm install

# Generate type stubs
npm run dev:prepare

# Develop with the playground
npm run dev

# Build the playground
npm run dev:build

# Run ESLint
npm run lint
# or npx eslint .

# Run Vitest
npm run test
npm run test:watch

# Release new version
npm run release
```

<!-- Badges -->

[npm-version-src]: https://img.shields.io/npm/v/seo/latest.svg?style=flat&colorA=18181B&colorB=28CF8D
[npm-version-href]: https://npmjs.com/package/seo
[npm-downloads-src]: https://img.shields.io/npm/dm/seo.svg?style=flat&colorA=18181B&colorB=28CF8D
[npm-downloads-href]: https://npmjs.com/package/seo
[license-src]: https://img.shields.io/npm/l/seo.svg?style=flat&colorA=18181B&colorB=28CF8D
[license-href]: https://npmjs.com/package/seo
[nuxt-src]: https://img.shields.io/badge/Nuxt-18181B?logo=nuxt.js
[nuxt-href]: https://nuxt.com


CHANGELOG.md
# Changelog

## v0.0.8

[compare changes](https://github.com/tubstrr/seo/compare/v0.0.7...v0.0.8)

## v0.0.7

[compare changes](https://github.com/tubstrr/seo/compare/v0.0.6...v0.0.7)

## v0.0.6

[compare changes](https://github.com/tubstrr/seo/compare/v0.0.5...v0.0.6)

## v0.0.5

[compare changes](https://github.com/tubstrr/seo/compare/v0.0.4...v0.0.5)

## v0.0.4

[compare changes](https://github.com/tubstrr/seo/compare/v0.0.3...v0.0.4)

## v0.0.3

[compare changes](https://github.com/tubstrr/seo/compare/v0.0.1...v0.0.3)

### 🏡 Chore

- Changed package name ([02cc665](https://github.com/tubstrr/seo/commit/02cc665))
- **release:** V0.0.2 ([9ddb8dc](https://github.com/tubstrr/seo/commit/9ddb8dc))

### ❤️ Contributors

- Tubstrr <me@jonknoll.dev>

## v0.0.2

[compare changes](https://github.com/tubstrr/seo/compare/v0.0.1...v0.0.2)

### 🏡 Chore

- Changed package name ([02cc665](https://github.com/tubstrr/seo/commit/02cc665))

### ❤️ Contributors

- Tubstrr <me@jonknoll.dev>

## v0.0.1


test/fixtures/basic/nuxt.config.ts
import seo from "../../../src/module";

export default defineNuxtConfig({
  modules: [seo],
});


src/module.ts
import {
  defineNuxtModule,
  addPlugin,
  createResolver,
  addImportsDir,
} from "@nuxt/kit";

// Module options TypeScript interface definition
export interface SeoModuleOptions {
  auto: boolean;
  locale: string;
  title?: string;
  description?: string;
  general: {
    tagline?: string;
    social?: Array<{
      type:
        | "facebook"
        | "twitter"
        | "x"
        | "youtube"
        | "instagram"
        | "linkedin"
        | "pinterest"
        | "tumblr"
        | "reddit"
        | "other";
      link: string;
    }>;
    defaultImage?: string;
  };
  organization?: {
    show: boolean;
    name: string;
    logo: string;
    url: string;
    type?: string;
    address?: {
      type: string;
      streetAddress: string;
      addressLocality: string;
      addressRegion: string;
      postalCode: string;
      addressCountry: string;
    };
    extra?: Array<{
      key: string;
      value: any;
    }>;
  };
  website?: {
    show: boolean;
    name: string;
    url: string;
    type?: string;
    search?: {
      show: boolean;
      action: string;
    };
    extra?: Array<{
      key: string;
      value: any;
    }>;
  };
  webpage?: {
    show: boolean;
    url?: string;
    image?: string;
    datePublished?: string;
    dateModified?: string;
    type?: string;
    extra?: Array<{
      key: string;
      value: any;
    }>;
  };
  breadcrumbs?: {
    show: boolean;
    excludedPaths?: Array<string>;
  };
  schemas?: Array<{
    hid: string;
    schema: object;
  }>;
  titles?: {
    separator?: string;
    template?: string;
    params?: object;
  };
  favicon?: {
    link?: Array<{
      rel:
        | "apple-touch-icon"
        | "icon"
        | "mask-icon"
        | "manifest"
        | "shortcut icon";
      href: string;
      sizes?: "16x16" | "32x32" | "96x96" | "192x192" | "512x512";
      color?: string;
    }>;
    meta?: Array<{
      name: "msapplication-TileColor" | "theme-color";
      content: string;
    }>;
  };
  link?: Array<{
    crossorigin?: string;
    href?: string;
    hreflang?: string;
    media?: string;
    referrerpolicy?: string;
    rel?: string;
    sizes?: string;
    title?: string;
  }>;
  meta?: Array<{
    name?: string;
    content?: string;
    charset?: string;
    "http-equiv"?: string;
  }>;
  script?: Array<{
    async?: boolean;
    charset?: string;
    crossorigin?: string;
    defer?: boolean;
    integrity?: string;
    nomodule?: boolean;
    nonce?: string;
    src?: string;
    type?: string;
    innerHTML?: string;
    children?: string;
  }>;
}

export default defineNuxtModule<SeoModuleOptions>({
  meta: {
    name: "seo",
    configKey: "seo",
  },
  // Default configuration options of the Nuxt module
  defaults: {
    auto: true,
    locale: "en_US",
    general: {
      tagline: "",
      social: [],
      defaultImage: "",
    },

    organization: {
      show: true,
      type: "Organization",
      name: "",
      logo: "",
      url: "",
      address: {
        type: "PostalAddress",
        streetAddress: "",
        addressLocality: "",
        addressRegion: "",
        postalCode: "",
        addressCountry: "",
      },
      extra: [],
    },

    website: {
      show: true,
      type: "WebSite",
      name: "",
      url: "",
      search: {
        show: false,
        action: "",
      },
      extra: [],
    },

    webpage: {
      show: true,
      type: "WebPage",
      url: "",
      image: "",
      datePublished: "",
      dateModified: "",
      extra: [],
    },

    breadcrumbs: {
      show: true,
      excludedPaths: [],
    },

    favicon: {
      link: [],
      meta: [],
    },

    titles: {
      separator: "-",
      template: "%s %separator %site.name",
      params: {},
    },
  },

  setup(options, nuxt) {
    const resolver = createResolver(import.meta.url);

    // Add default title params
    // Failsafe for missing options
    options.titles = options.titles || {};
    options.titles.params = {
      site: {
        name: options?.website?.name,
      },
      separator: options.titles.separator,
      tagline: options.general.tagline,
    };

    // Add seo key to app.seo config
    nuxt.options.runtimeConfig.app.seo = options;

    // Add composables directory
    addImportsDir(resolver.resolve("./runtime/composables"));

    // TODO: Add autoRun plugin if auto is true
    // if (options.auto) {
    addPlugin(resolver.resolve("./runtime/autoRun"));
    // }
  },
});


playground/server/api/seo.ts
export default defineEventHandler((event) => {
  const { url } = event.node.req;
  const searchParams = new URLSearchParams(url?.split("?")?.[1]);
  const path = searchParams.get("path");

  if (!path) {
    return {
      error: "No path provided",
    };
  }

  const dataByRoute = {
    "/": {
      title: "Home",
      description: "The home page",
      webpage: {
        image: "https://jonknoll.dev/wp-content/uploads/2021/03/use-2.jpg",
        datePublished: "2022-01-01",
        dateModified: "2022-01-02",
      },
    },
    "/about": {
      title: "About Us Page",
      description: "The about page",
      schemas: [
        {
          hid: "justinBieber",
          schema: {
            "@context": "https://schema.org",
            "@type": "Event",
            name: "Name",
            description:
              "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Cras vestibulum nulla ut nibh vulputate egestas.",
            image: "https://jonknoll.dev/wp-content/uploads/2021/03/use-2.jpg",
            startDate: "2023-12-08T08:00",
            endDate: "2023-12-08T12:00",
            eventStatus: "https://schema.org/EventScheduled",
            eventAttendanceMode:
              "https://schema.org/OfflineEventAttendanceMode",
            location: {
              "@type": "Place",
              name: "Mess",
              address: {
                "@type": "PostalAddress",
                streetAddress: "1824 W Grand Ave #200",
                addressLocality: "Chicago",
                addressRegion: "IL",
                postalCode: "60622",
                addressCountry: "US",
              },
            },
            performer: {
              "@type": "MusicGroup",
              name: "Justin Bieber",
            },
            offers: {
              "@type": "Offer",
              name: "General Admission",
              price: "50",
              priceCurrency: "",
              validFrom: "2023-12-04",
              url: "https://thisismess.com/",
              availability: "https://schema.org/InStock",
            },
          },
        },
      ],
    },
    "/about/our-history": {
      title: "Our Great History",
      description: "The contact page",
      webpage: {
        datePublished: "2022-03-01",
        dateModified: "2022-03-02",
      },
    },
  };

  return dataByRoute[path];
});


playground/nuxt.config.ts
export default defineNuxtConfig({
  seo: {
    auto: false,
    locale: "en_US",
    general: {
      tagline: "We are a fiercely independent design and development studio.",
      defaultImage:
        "https://s3.amazonaws.com/thisismess-assets/mess/live/static/img/og-share-image.jpg",
      social: [
        {
          type: "facebook",
          link: "https://www.facebook.com/thisismess/",
        },
        {
          type: "twitter",
          link: "https://twitter.com/thisismess",
        },
        {
          type: "instagram",
          link: "https://www.instagram.com/thisismess",
        },
        {
          type: "linkedin",
          link: "https://www.linkedin.com/company/mess-marketing",
        },
        {
          type: "youtube",
          link: "https://www.youtube.com/@MessMakesVideos",
        },
      ],
    },
    organization: {
      name: "This is Mess",
      url: "https://thisismess.com",
      logo: "https://s3.amazonaws.com/thisismess-assets/mess/live/static/img/favicons/apple-touch-icon.png",
      address: {
        streetAddress: "1824 W Grand Ave #200",
        addressLocality: "Chicago",
        addressRegion: "IL",
        postalCode: "60622",
        addressCountry: "US",
      },
      extra: [
        {
          key: "ContactPoint",
          value: {
            name: "Jon Knoll",
            url: "https://jonknoll.dev",
            email: "jknoll@thisismess.com",
            image: "https://jonknoll.dev/wp-content/uploads/2021/03/use-2.jpg",
          },
        },
        { key: "email", value: "contact@thisismess.com" },
      ],
    },
    website: {
      show: true,
      name: "Mess Marketing",
      url: "https://thisismess.com",
      search: {
        show: true,
        action: "https://thisismess.com/search?search=",
      },
    },
    favicon: {
      link: [
        {
          rel: "apple-touch-icon",
          sizes: "180x180",
          href: "https://s3.amazonaws.com/thisismess-assets/mess/live/static/img/favicons/apple-touch-icon.png",
        },
        {
          rel: "icon",
          sizes: "32x32",
          href: "https://s3.amazonaws.com/thisismess-assets/mess/live/static/img/favicons/favicon-32x32.png",
        },
        {
          rel: "icon",
          sizes: "16x16",
          href: "https://s3.amazonaws.com/thisismess-assets/mess/live/static/img/favicons/favicon-16x16.png",
        },
        {
          rel: "manifest",
          href: "https://s3.amazonaws.com/thisismess-assets/mess/live/static/img/favicons/site.webmanifest",
        },
        {
          rel: "mask-icon",
          color: "#37dbff",
          href: "https://s3.amazonaws.com/thisismess-assets/mess/live/static/img/favicons/safari-pinned-tab.svg",
        },
        {
          rel: "shortcut icon",
          href: "https://s3.amazonaws.com/thisismess-assets/mess/live/static/img/favicons/favicon.ico",
        },
      ],
      meta: [
        {
          name: "msapplication-TileColor",
          content: "#ffffff",
        },
        {
          name: "theme-color",
          content: "#ffffff",
        },
      ],
    },
  },
  modules: ["../src/module"],
  devtools: { enabled: true },
});
